Adding support for postgresql as datastore for customers service

nimrodshn opened the pull request:
cc: @jhernand @moolitayer 

TODO:

- [X] Add script file to create the database and table.

jhernand commented on cmd/customers-service/psql_customers_service.go:
```diff
@@ -0,0 +1,137 @@
+	db *sql.DB
+}
+
+// NewPsqlCustomersService is a constructor for the EtcdCustomersService struct.
```
Replace `Etc` with SQL, here and in other places.

jhernand commented on cmd/customers-service/psql_customers_service.go:
```diff
@@ -0,0 +1,137 @@
+// NewPsqlCustomersService is a constructor for the EtcdCustomersService struct.
+func NewPsqlCustomersService(user, password, dbname, host, port, sslmode string) (*PsqlCustomersService, error) {
+	connStr := fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=%s",
+		host, port, user, password, dbname, sslmode)
```
I may be better to avoid this long collection of parameters, and pass directly the connection string.

jhernand commented on cmd/customers-service/psql_customers_service.go:
```diff
@@ -0,0 +1,137 @@
+}
+
+// Close closes the etcd customers service client.
+func (service *PsqlCustomersService) Close() {
```
This should return `error`, so that it implements the `io.Closer` interface.

nimrodshn replied:
I will add it to both implementations in a separate PR.

jhernand commented on cmd/customers-service/psql_customers_service.go:
```diff
@@ -0,0 +1,137 @@
+		result.OwnedClusters = customer.OwnedClusters
+	}
+
+	_, err = service.db.Exec("INSERT into customers (id, name, owned_clusters) VALUES ($1, $2, $3)",
```
This is a simple statement, so it may make no sense to format it better, but in general I'd suggest to try to use multiple lines:

```go
_, err = service.db.Exec(`
    insert into customers (
        id,
        name,
        owned_clusters
    ) values (
        $1,
        $2,
        $3
    )
')
```

This way it is easier to read, and it is also easier to add/modify/remove/reorder columns.

I'd also suggest to write the SQL keywords lower-case, it is also more readable.

I'd also try to avoid the use of `pq.Array` and have instead a more normalized schema, with its own `owned_clusters` table.

nimrodshn replied:
@jhernand  Not sure I understand, also I'm not sure how to avoid the use of `pq.Array` what do you suggest?

jhernand replied:
I suggest to have a `customers` table:

```sql
create table customers (
  id text primary key,
  name text
);
```

And another `owned_clusters` table to store the relationship between customers and the clusters they own:

```sql
create table owned_clusters (
  customer_id text references customers (id),
  cluster_id text unique,
);
```

Then, in order to insert a customers, you will need first to insert in the `customers` table and then in the `owned_clusters` table.

jhernand commented on cmd/customers-service/psql_customers_service.go:
```diff
@@ -0,0 +1,137 @@
+// List retrieves a list of current customers stored in datastore.
+func (service *PsqlCustomersService) List(args *ListArguments) (*CustomersList, error) {
+	var result *CustomersList
+	var items = make([]*Customer, 0)
```
You could maybe delay the allocation of this slice till you know the size it will have. Then you can allocate it for exactly the number of items it will have, instead of letting the Go runtime resize and reallocate it multiple times:

```go
if args != nil {
    args = make([]*Customer, args.Size)
    ...
} else {
    ...
}
```

jhernand commented on cmd/customers-service/psql_customers_service.go:
```diff
@@ -0,0 +1,137 @@
+	if args != nil {
+		rows, err = service.db.Query("SELECT * FROM customers ORDER BY id LIMIT $1 OFFSET $2", args.Size, args.Page*args.Size)
+	} else {
+		rows, err = service.db.Query("SELECT * FROM customers")
```
Not sure if we should allow this. What if there are 10000000 customers? Will we allocate and return an array of that size? I think we could set a default value for `size`, something like 100. And even maybe a max value, something like 1000.

jhernand commented on cmd/customers-service/psql_customers_service.go:
```diff
@@ -0,0 +1,137 @@
+}
+
+func (service *PsqlCustomersService) deleteAll() error {
+	_, err := service.db.Exec("DELETE FROM customers")
```
This is an extremely dangerous operation, not sure if we should have it at all. I understand that it is only intended for tests, but even for that I am not sure if it is OK to have it.

nimrodshn commented:
@jhernand PTAL.

One comment you should note - in `List` there was no way to allocate correctly the amount of items **even if the `Size` list argument is supplied ** - there is really no way to know the result size - e.g. the client is asking the last page from our customer list and is supplying a size limit of 100, the actual page might be a lot smaller (Lets say the list is of 304 items, so the last page is of size 4!) so allocating 100 pointers is not correct. 
I can't think of a smarter way than to allocate an empty slice and append.

jhernand commented:
My understanding is that the Go slice growing algorithm doubles the capacity of the slice when there is no space, and then it copies the old content. So, say that you the page size is 100, and that you start with an slice of capacity 0, like this:

```go
items := make([]*Item, 0)
```

For the first append you will be allocating an array of capacity 1. For the second you will be allocating an array of capacity 2, and copying 1 item. For the third you will be allocating an array of capacity 4, and copying 2 items. For the fourth there is no need to reallocate, there is still room. So on. In total you will be doing the following allocations:

```
iteration:allocation
0:1
1:2
2:4
4:8
8:16
16:32
32:64
64:128
```

So in order to add those 100 items you are doing 8 allocations, 127 copies and wasting 28 slots.

On the other hand, if you initialize the array with a capacity that you know will be enough:

```go
items := make([]*Item, 0, 100)
```

Then you do one allocation, 0 copies, and no waste.

In your example with 304 items, these savings apply to the first three pages. For the last page it is true that you are wasting 96 slots.

All in all, it is not that relevant, in general, we have fast computers and lots of memory, and the slow path is the network, so there won't be much difference. But it is so easy to make it better.

jhernand commented on cmd/customers-service/customers.sql:
```diff
@@ -0,0 +1,10 @@
+create database customers;
+create table customers (
+  id             text not null unique primary key,
+  name           varchar(15) not null,
```
You can use `text` here. PostgreSQL manages well unlimited size fields, so no need to impose an artificial limitation.

jhernand commented on cmd/customers-service/customers.sql:
```diff
@@ -0,0 +1,10 @@
+create table customers (
+  id             text not null unique primary key,
+  name           varchar(15) not null,
+  owned_clusters text ARRAY
```
I think this isn't needed; you are storing the identifiers of the owned clusters in the `owned_clusters` table.

nimrodshn replied:
Shoot I forgot to delete this :cry:

jhernand commented on cmd/customers-service/sql_customers_service.go:
```diff
@@ -0,0 +1,172 @@
+	var result Customer
+
+	// Get the customer infromation
+	rows, err := service.db.Query("select * from customers where id=$1", id)
```
If we use `select *` then the order of the columns may change unexpectedly, and then you may be swapping the `id` and `name` columns. I think we should explicitly indicate the columns: `select id, name ...`. And probably you only want to retrieve the `name`, as you already know the `id`.

jhernand commented on cmd/customers-service/sql_customers_service.go:
```diff
@@ -0,0 +1,172 @@
+	}
+
+	for rows.Next() {
+		if err = rows.Scan(&result.ID, &result.Name); err != nil {
```
You already know the `id`, so you could scan just the name.

jhernand commented on cmd/customers-service/sql_customers_service.go:
```diff
@@ -0,0 +1,172 @@
+	}
+
+	// Retrieve customers owned clusters.
+	for _, customer := range items {
```
Note that this has the typical N+1 problem: you are sending to the database as many queries as customers. That is usually very slow. Consider using a single query to retrieve the customers with their clusters:

```sql
select
  customers.id,
  customers.name,
  owned_clusters.cluster_id
from
  customers
join
  owned_clusters on owned_clusters.customer_id = customers.id
where
  customers.id = 123
```

That has the disadvantage that the customer id and name are replicated multiple times in the result set.

Another alternative is to send one query to get the customers, and then another query to get the owned clusters, but only for that particular set of customers:

```sql
select
  customer_id,
  cluster_id
from
  owned_clusters
where
  cluster_id in (123, 456, 789, ...)
```

jhernand commented on cmd/customers-service/sql_customers_service.go:
```diff
@@ -0,0 +1,172 @@
+
+	result = &CustomersList{
+		Items: items,
+		Page:  1,
```
Should this be `args.Page`?

jhernand commented on cmd/customers-service/sql_customers_service.go:
```diff
@@ -0,0 +1,172 @@
+		Items: items,
+		Page:  1,
+		Size:  int64(len(items)),
+		Total: int64(len(items)),
```
Total should be the total number of items that match the search criteria, ignoring the paging. To calculate it you will have to send another query:

```sql
select
  count(*)
from
  customers
where
  ...
```

nimrodshn replied:
Yeah I just thought of that, thanks :+1:

jhernand commented on cmd/customers-service/sql_customers_service.go:
```diff
@@ -0,0 +1,172 @@
+}
+
+// List retrieves a list of current customers stored in datastore.
+func (service *SQLCustomersService) List(args *ListArguments) (*CustomersList, error) {
```
All the SQL statements performed by this method should be within a transaction, otherwise you may be returning inconsistent results.

cben replied:
I'm not certain db transactions (at least as with *typical* DB settings) isolate you from seeing concurrent changes mid-way.
IIRC they only isolate *others* from seeing *partial* changes by you — they guarantee atomic commit or rollback (?)
I might be speaking FUD though :)
Reading material (I only skimmed some of those):

- https://www.postgresql.org/docs/9.6/static/transaction-iso.html
- https://blog.acolyer.org/2016/02/24/a-critique-of-ansi-sql-isolation-levels/
- https://www.infoq.com/news/2016/03/real-world-consistency
  > This means that you can and will experience anomalies in production environments, even if you use ACID transactions, but most developers are not aware of that.

nimrodshn commented:
@jhernand PTAL again. :smile:

jhernand commented on cmd/customers-service/sql_customers_service.go:
```diff
@@ -0,0 +1,234 @@
+		result.OwnedClusters = customer.OwnedClusters
+	}
+
+	_, err = service.db.Exec(`insert into customers (
```
I think this looks/reads better if you put the `insert ...` in the next line, like you did in the next insert.

jhernand commented on cmd/customers-service/sql_customers_service.go:
```diff
@@ -0,0 +1,234 @@
+			$1,
+			$2
+		)`,
+		result.ID, result.Name)
```
I'd also suggest to put the values for $1 and $2 in separate lines.

jhernand commented on cmd/customers-service/sql_customers_service.go:
```diff
@@ -0,0 +1,234 @@
+	// Get the customer infromation
+	// If not customer found return nil pointer and nil error.
+	// (See customers_service.go for more details)
+	rows, err := service.db.Query(`select (name) from customers where id=$1`, id)
```
Why the parenthesis around `name`?

jhernand commented on cmd/customers-service/sql_customers_service.go:
```diff
@@ -0,0 +1,234 @@
+	customersToClusters := make(map[string][]string)
+	rows, err = service.db.Query(`select customer_id, cluster_id from owned_clusters
+		where customer_id in (
+			select id from customers limit $1 offset $2
```
This may work today, because there is no search criteria, but it will fail in unexpected ways when a search criteria is added.

Instead of doing this nested query my suggestion is to build an string that looks like this:

```
('123', '456', ...)
```

Where 123, 456, etc, are the customer identifiers that you retrieved in the previous step. Then you can build the query, using that string instead of using the nested query.

nimrodshn replied:
@jhernand Can you use variadic parameter in an SQL query? I could not find how to write the query.. :sob:

nimrodshn commented:
@jhernand PTAL - I ended up using the JOIN approach.

jhernand commented on cmd/customers-service/sql_customers_service.go:
```diff
@@ -0,0 +1,243 @@
+		 join owned_clusters on
+		(owned_clusters.customer_id = customers.id)
+		where id in ($1)`,
+		strings.Join(ids, " "))
```
Does this work? I think you will need to add single quotes around each id.

nimrodshn replied:
nope I think you're confusing split with join. (See [here](https://play.golang.org/p/upwox_kHfBB).)

nimrodshn replied:
Also, I wasn't sure - should I add transactions to all methods?

jhernand replied:
I am almost sure that this is not going to work as you expect. For example, if the identifiers are `abc` and `cde` what this will do is create a sting `abc cde`, and then it will generate a query like this:

```sql
select ... where id in ('abc cde') ...
```

That is not what you want. In the best case it will return no results. In the worst case, if there is actually a customer with id `abc cde`, it will return wrong results.

What you want is a query like this:

```sql
select ... where id in ('abc', 'cde') ...
```

To make this work you need to build the SQL yourself, something like this:

```go
// First build the set of quoted ids:
ids = ...
quotedIds := make([]string, len(ids))
for i, id := range ids {
  quotedIds[i] = fmt.Sprintf("'%s'", id)
}
idSet := strings.Join(quotedIds, ", ")

// The build the SQL statement:
sqlStmt := fmt.Sprintf("select ... where id in (%s) ...", idSet)

// Now run the SQL statement:
... := db.Query(sqlStmt)
```

nimrodshn replied:
@jhernand I have just changed it! to you're method

jhernand commented on cmd/customers-service/sql_customers_service_test.go:
```diff
@@ -0,0 +1,204 @@
+var service *SQLCustomersService
+var localTesting = true
+
+func TestMain(m *testing.M) {
```
These tests will fail if there is no such database available, so it should be disabled by default.

nimrodshn replied:
Done. PTAL.

nimrodshn replied:
@jhernand

nimrodshn commented:
@jhernand For youre second comment, I new there would be @yaacov pr that changes the entire cli so I didnt touch on starting it till that got merged plus I prefer to keep pr's small, and this one is a bit large and complex as is don't you think?

nimrodshn commented:
@jhernand **Please dont merge this yet I am trying to make this query as safe as possible.**

nimrodshn commented:
@jhernand So my concern was that using this query we are exposing ourselves to SQL injections but since the `ids` are server generated the query seems OK. @cben @moolitayer WDYT?  (we are talking about [this](https://github.com/container-mgmt/dedicated-portal/pull/39/files#diff-9797f0b4e6530b178bc82c2751446ad6R206) query)

nimrodshn commented:
> I have a question regarding the way the set of ids is generated, see inside.
Refaco
> 
> In addition I miss a few things:
> 
> Transactions.
> Where is this service created?
> What happens with the tests if there is no database running?
> I think we can merge one the question about the generation of the set of ids is addressed. The other things can go in separate pull requests, in my opinion.

jhernand commented on cmd/customers-service/sql_customers_service.go:
```diff
@@ -0,0 +1,253 @@
+
+// Add adds a single customer to psql database.
+func (service *SQLCustomersService) Add(customer Customer) (*Customer, error) {
+	tx, err := service.db.Begin()
```
What if there is an error and you return before committing the transaction? I think you will leave it open. I'd suggest you do something like this:

```go
func (service *SQLCustomersService) Transaction(work func() error) error {
    tx, err := service.db.Begin()
    if err != nil {
        return err
    }
    err = work()
    if err != nil {
        rollbackErr := tx.Rollback()
        if rolllbackErr != nil {
           return fmt.Errorf(
               "can't rollback after error: %s: %s",
               err,
               rollbackErr,
        }
        return err
    }
    return tx.Commit()
}
```

Then, from your function you can do something like this:

```go
func (service *SQLCustomersService) Add(customer Customer) (customer *Customer, err error) {
    err = service.Transaction(func () error {
        // The body of your function
    })
    return
}
```

If it is OK if you want to do this in a separate pull request.

cben commented on cmd/customers-service/sql_customers_service.go:
```diff
@@ -0,0 +1,235 @@
+		 join owned_clusters on
+		(owned_clusters.customer_id = customers.id)
+		where id in ($1)`,
+		strings.Join(ids, `","`))
```
I suspect that's still wrong.  IIUC `Query()` quotes once per placeholder, so if strings is ["foo", "bar"] this will give
`select ... where id in ('foo,bar')`
But you want to quote each string separately:
`select ... where id in ('foo', 'bar')`

- https://stackoverflow.com/a/20275714/239657 gives messy approach with dynamically generated query *template* that include variable number of placeholders.
- With pq driver, https://stackoverflow.com/a/38878826/239657 says we can pass a `pq.Array` and have all elements quoted for us.

nimrodshn replied:
@jhernand wanted to avoid the use of `pq.Array` so that this solution would not be *sepcific* to `Postgres` since.

nimrodshn commented:
@jhernand @cben  I've fixed the `List` query so that it works now as well as added tests to check that the `ownedClusters` returns correctly, in addition I have fixed the allocation of `Items` field so that it returns better results now.
